lib/logstash/inputs/beats/message_listener.rb

```diff
@@ -119,6 +122,35 @@ def flush_buffer(ctx)
       end
     end
 
+    def extract_tls_peer(hash)
+      if @input.client_authentication_metadata?
+        tls_session = ctx.channel().pipeline().get("ssl-handler").engine().getSession()
+        tls_verified = false
+
+        # throws SSLPeerUnverifiedException if unverified
+        begin
+          tls_session.getPeerCertificates()
+          tls_verified = true
+        rescue SSLPeerUnverifiedException => e
```

Comment:
In JRuby exceptions are very expensive, is there an alternative way to check if a peer certificate was presented ?

Reply:
In theory, yes. Netty should raise a specific `HandshakeCompletionEvent` event in https://github.com/logstash-plugins/logstash-input-beats/blob/master/src/main/java/org/logstash/beats/ConnectionHandler.java#L64 which the certificate data can be extracted in the very same way.

The problem is that I don't know how to assign the data to the event as there are no obvious references to the event object at that method.

Reply:
I would say that this is an edge case anyway. It means that the client certificate metadata is being asked while allowing to _not_ use a client certificate. I would say that this would be used more for testing then anything else.

BTW `ssl_verify_mode => "peer"` does not even works and this was never reported. See https://github.com/logstash-plugins/logstash-input-beats/issues/326

Reply:
Maybe there could be an optimization of checking `force_peer` and then never call this `getPeerCertificates()` method (I am not really sure what happens under the covers)